user-console/src/main/java/org/pentaho/mantle/client/commands/NewDropdownCommand.java

```diff
@@ -126,6 +131,19 @@ protected void onPreviewNativeEvent( final NativePreviewEvent event ) {
             if ( (char) nativeEvent.getKeyCode() == KeyCodes.KEY_ESCAPE ) {
               event.cancel();
               hide();
+            } else if ( (char) nativeEvent.getKeyCode() == KeyCodes.KEY_TAB ) {
+              Element element = DOM.eventGetTarget( nativeEvent );
+              if( element.getParentElement().getParentElement().getNextSiblingElement() == null ) {
+                event.cancel();
+                hide();
+              }
+            } else if ( nativeEvent.getShiftKey() && (char) nativeEvent.getKeyCode() == KeyCodes.KEY_TAB ) {
+              Element element = DOM.eventGetTarget( nativeEvent );
+              if( element.getParentElement().getParentElement().getPreviousSiblingElement() == null ) {
+                event.cancel();
+                hide();
+                element.getParentElement().getParentElement().getParentElement().focus();
+              }
```

Comment:
This code snippet will not be able to fetch the button index using the element. So, this way is not suitable.

Reply:
Hi @addagudi, you're right, that's what happens when writing code on the github reply box...

I believe I have found a way at it, though.

First, define the following function in the `ElementUtils` class (in the pentaho-commons-gwt project):
```java
public class ElementUtils {
/**
   * Gets the widget which is listening to events on the given root element.
   * <p>
   *   Widgets are associated with their root element on attachment (see <code>Widget#onAttach()</code>).
   * </p>
   * <p>
   *   Based on https://stackoverflow.com/a/17863305/178749.
   * </p>
   * @param element The widget's root element.
   * @return The widget, if one is associated with the <code>element</code>;
   *         <code>null</code>, otherwise.
   */
  public static Widget getWidgetOfRootElement( Element element ) {
    EventListener listener = DOM.getEventListener( element );
    return ( listener instanceof Widget ) ?  (Widget) listener : null;
  }
}
```

Then use it in the above proposed code:
```java
} else if ( (char) nativeEvent.getKeyCode() == KeyCodes.KEY_TAB ) {
  Element buttonElement = DOM.eventGetTarget( nativeEvent );
  Widget button = ElementUtils.getWidgetOfRootElement( buttonElement );
  if ( button != null ) {
    // Need to store the buttonPanel in class property.
    int buttonIndex = buttonPanel.getWidgetIndex( button );
    int goToButtonIndex = buttonIndex + (nativeEvent.getShiftKey() ? -1 : 1);
    if ( goToButtonIndex >= 0 && goToButtonIndex < buttonPanel.getWidgetCount() ) {
      buttonPanel.getWidget( buttonIndex ).setFocus( true );
    }
  }
```